### PR TITLE
include self link in schema

### PR DESCRIPTION
**Pull request by NobodysNightmare**
Fixes test failure of https://community.openproject.org/work_packages/18570
## Notes
- self link will only be displayed outside of a form, inside a form there is additional information that is not reachable via a link
- `_type` indicating `Schema` is still missing... this goes down to a flaw in the specification, because we want two things:
  - indicate that there is a field called `_type` on the thing described by the schema
  - indicate that the schema has a `_type` of `Schema`

With the current schema layout it is not possible to do both, because that would result in a duplicate key called `_type`. We therefore only indicate the former. At a later point the specification on this needs to be revised.

I would really apprechiate if you could **merge #2550 _before_ merging this PR**. This would move any conflicts into the smaller PR.

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://nobodysnightmare-fix-self-link-schema.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://nobodysnightmare-fix-self-link-schema.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

**Comment by ulferts** [comment]
@NobodysNightmare needs rebasing, sorry.

**Comment by ulferts** [comment]
This is broken and needs rebasing @NobodysNightmare

File: doc/apiv3-documentation.apib [hunk 1]
```diff
@@ -1608,7 +1608,6 @@ Note that regular expressions used in the API follow the rules of [Ruby Regular
     + Body
 
             {
-                "_type": "Schema",
```
Comment (ulferts) [inline doc/apiv3-documentation.apib:1611]: Shouldn't we write what you wrote as a preamble to this PR into the documentation as well?

File: lib/api/v3/work_packages/schema/work_package_schema.rb [hunk 2]
```diff
@@ -63,19 +59,19 @@ def assignable_statuses_for(user)
           end
 
           def assignable_versions
-            @work_package.assignable_versions if defines_assignable_values?
+            @work_package.assignable_versions unless @work_package.nil?
           end
 
           def assignable_priorities
-            @work_package.assignable_priorities if defines_assignable_values?
+            IssuePriority.where(active: true)
```
Comment (ulferts) [inline lib/api/v3/work_packages/schema/work_package_schema.rb]: Note to self and whoever reads this. It would be better to have an `active` scope on IssuePriority to avoid AR methods here. I am aware, that the code was just moved, though.

File: lib/api/v3/work_packages/schema/work_package_schema.rb [hunk 3]
```diff
@@ -63,19 +59,19 @@ def assignable_statuses_for(user)
           end
 
           def assignable_versions
-            @work_package.assignable_versions if defines_assignable_values?
+            @work_package.assignable_versions unless @work_package.nil?
```
Comment (ulferts) [inline lib/api/v3/work_packages/schema/work_package_schema.rb]: For patterns like this you can use:

`@work_package.try(:assignable_versions)`

File: lib/api/v3/work_packages/schema/work_package_schema.rb [hunk 4]
```diff
@@ -63,19 +59,19 @@ def assignable_statuses_for(user)
...
           end
 
           def assignable_categories
-            @work_package.assignable_categories if defines_assignable_values?
+            project.categories
           end
 
           def available_custom_fields
-            @project.all_work_package_custom_fields & @type.custom_fields.all
+            project.all_work_package_custom_fields & type.custom_fields.all
```
Comment (ulferts) [inline lib/api/v3/work_packages/schema/work_package_schema.rb:84]: @NobodysNightmare will this conflict with what you are doing in the other PR right now? Any preference of merging order?

**Comment by ulferts** [comment]
@NobodysNightmare does what it says and I could spot no flaw.

I'd still like to discuss whether it is correct to have the schema **embedded** inside the work package form, thereby suggesting that it is a resource it it's own right, and then not having a self link to said resource. The standard is not my strong suit but I think that we violate it here and I am wondering whether we can simply move the schema to anything other than the embedded section. Will that make us  good again?

File: doc/apiv3-documentation.apib (hunk 1 above)
Comment (NobodysNightmare) [inline doc/apiv3-documentation.apib:1611]: I'd like to update the documentation as soon as we have the real solution to be honest...

As nobody seems to be interested in requiring the `_type` for `PATCH` requests I'd say lets change that now:
1. I reintroduce the `"_type": "Schema"`
2. I remove `_type` attributes from schema listings
3. Specification is updated to state that the `_type` is not required for `PATCH`ing

Any objections?

File: lib/api/v3/work_packages/schema/work_package_schema.rb (hunk 2 above)
Comment (NobodysNightmare) [inline lib/api/v3/work_packages/schema/work_package_schema.rb]: How hard is it to add this scope?

Should I just do it?

File: lib/api/v3/work_packages/schema/work_package_schema.rb (hunk 4 above)
Comment (NobodysNightmare) [inline lib/api/v3/work_packages/schema/work_package_schema.rb:84]: Git will tell us if it is conflicting :)

Merge whatever is ready first.

**Comment by NobodysNightmare** [comment]
@ulferts regarding standards conformity:

It might sound crazy, but I am also not pedantic enough to see where we are violating which standard. However, I will try my best.

What I have seen so far in the wild, and what we are also doing is that it is possible to have stuff embedded, that is not linkable. e.g. looking at our collections they will always have `_embedded/elements`, though there is no link for those elements except the collection itself (which only carries those elements as embedded).

Looking at the form we already have the `payload`, which is also **only** embedded. It has no self link and no self link would be valid, because the payload is an imaginary object that only becomes real (as in "has a link") as soon as you call the `PATCH` action.

**tl; dr:** I don't know what the HAL-police will say, but looking at the rest of _our_ API the embedded schema seems to be fitting (the rest is at least on the same degree of wrong-ish-ness).

File: lib/api/v3/work_packages/schema/work_package_schema.rb (hunk 2 above)
Comment (NobodysNightmare) [inline lib/api/v3/work_packages/schema/work_package_schema.rb]: Is it more than just `scope :active, where(active: true)`?

**Comment by NobodysNightmare** [comment]
I changed the specification to not require specifying a `_type` in `PATCH` anymore. Therefore I could make the schema more compliant with all other resources.

old schema:

``` JSON
{
  "_type": {
    "type": "MetaType",
    "...": "..."
  },
  "subject": {
    "type": "String",
    "...": "..."
  },
}
```

new schema:

``` JSON
{
  "_type": "Schema",
  "subject": {
    "type": "String",
    "...": "..."
  },
}
```

This means that a schema resource now should look more similar to all the other resources out there.

@0xF013 I don't know if this is a thing for you, but you might want to know.

**Comment by NobodysNightmare** [comment]
Go home @houndci, you are drunk!

**Comment by ulferts** [comment]
@NobodysNightmare needs rebasing, too.
